NIFI-1295: Adding UI controls for terminating threads

mcgilman opened the pull request:
NIFI-1295:
- Adding UI controls for terminating hung threads.
- Showing current number of terminated threads.

moranr commented:
@mcgilman , I was thinking about some suggestions to the formatting and tooltip labeling used on these. I think we're using the '/' (forward slash) inconsistently which could be a bit confusing.

How about formatting the thread count in the global status bar like so:
**3 (1)**

And the tooltip could read something like:
**Threads: Active (Terminated)**

Then when the thread icon shows on a processor, the count would again show as '# (#)' and the tooltip could be something like:
**3 active threads (1 terminated)**

I'm also thinking change the current queue/data size global status format to:
**_Object Count (Data Size)_**
i.e., use parentheses instead of the forward slash

This is consistent with the In & Out stats on components.

And one last thing I thought is adding the term 'threads' to the Terminate context menu action. So, change _Terminate_ to _Terminate threads_

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java:
```diff
@@ -3846,6 +3851,73 @@ public QueueSize getTotalFlowFileCount(final ProcessGroup group) {
         return new QueueSize(count, contentSize);
     }
 
+    public class GroupStatusCounts {
```
This is nit-picky but given that this class can be declared as static, since it doesn't reference FlowController private variables/methods, and given that FlowController is already nearly 4,000 lines of code, I'd recommend pulling this class out into its own .java file instead of making it an inner class of FlowController.

markap14 replied:
And honestly, at that point, you could even get rid of the getGroupStatusCounts(ProcessGroup) method all together if you wanted, as you could just use the FlowController to get the root group and then just pass that root group to the constructor of GroupStatusCounts...

mcgilman replied:
This cannot be static as it needs access to the `processScheduler`

markap14 replied:
Ah yes, I overlooked that. Is a bit difficult to find these things in a Github Diff page :)

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/StandardProcessSession.java:
```diff
@@ -330,7 +330,7 @@ public void checkpoint() {
 
     @Override
-    public void commit() {
+    public synchronized void commit() {
```
Good - i think i made a mistake in missing this in my last refactor.

markap14 commented:
+1 to @moranr's suggestion for formatting, also.

markap14 commented:
After testing, all looks good to me. I'm a +1 to merge if you want to make those format changes suggested by Rob and optionally the class refactoring that I suggested in the FlowController. Thanks! This is definitely an awesome feature to have added in!

mcgilman commented:
@moranr @scottyaslan @markap14 I've addressed the PR feedback above. Please have a look at my most recent commit. A couple notes:

- I also found and addressed an issue when replicating the terminate threads request throughout the cluster.
- I left the Terminate content menu alone since it replaces the Stop option. Like Start/Stop, I think simply stating the action is sufficient.
- I did not address the formatting of the queue stats. When investigating I discovered an issue with this formatting. It turns out that it is already formatted in the proposed way when in clustered mode. In standalone this formatting is different. Unfortunately, its a bigger change then I'd like to make in this PR. As a result, I filed a new JIRA [1] to address it in the future.

[1] https://issues.apache.org/jira/browse/NIFI-5067

moranr commented:
Thanks @mcgilman, looks good. Could go either way on the Terminate v Terminate threads menu, so unless anyone has a strong opinion for adding "threads," I understand your logic for not making the change. Good catch on the cluster formatting; the formatting changes made here so far are a good step.

scottyaslan commented:
Looks good!

markap14 commented:
@mcgilman this is a great feature that I think will make a lot of people's lives easier. Thanks for updating the UI to support this! Also thanks to @scottyaslan and @moranr for your feedback and testing. Everything looks good at this point, so I've merged to master.
